image_masking.py

```diff
@@ -0,0 +1,125 @@
+                    corners[cornerIndex + 1] - corners[cornerIndex + 3 * 2 + 1]
+                }; 
+
+                int2 rightVec = {
+                    corners[cornerIndex + 2 * 2] - corners[cornerIndex + 1 * 2],
+                    corners[cornerIndex + 2 * 2 + 1] - corners[cornerIndex + 1 * 2 + 1]
+                };
+
+                // calculate cross product for each side of array
+                int crossTop = topVec.x * (w - corners[cornerIndex + 1]) - topVec.y * (h - corners[cornerIndex]);
+                int crossRight = rightVec.x * (w - corners[cornerIndex + 3]) - rightVec.y * (h - corners[cornerIndex + 2]);
+                int crossBottom = bottomVec.x * (w - corners[cornerIndex + 5]) - bottomVec.y * (h - corners[cornerIndex + 4]);
+                int crossLeft = leftVec.x * (w - corners[cornerIndex + 7]) - leftVec.y * (h - corners[cornerIndex + 6]);
+
+                // our point is inside as long as every cross product is greater or equal to 0
+                bool inside = crossTop >= 0 && crossRight >= 0 && crossBottom >= 0 && crossLeft >= 0;
+
+                mask = inside ? fillValue : originalMask;
+            ''',
+            name='bbox_to_mask',
+        )
+
+        height, width, channels = base_mask.shape
+        mask = create_mask_kernel(base_mask, corners, height, width, self.fill_value)
+        return mask
+        
+    def mask_array(self, array, corners, xp):
+        """
+        Create a mask on a given array, using the OOB specified by corners
+
+        :param array: An array with three dimensions. The first dimension is the width of the feature map or image.
+        The second dimension is the height of the feature map or image. The third dimension is the batch dimension
+        (if you want to mask a 4 dimensional array, you'll first need to collapse the batch and channel dimension, but
+        make sure to also pad the given corners!, that means increase the number of corners to match
+        batch size * number_of_channels).
+
+        :param corners: An array with three dimensions. The first dimension is the batch axis (similar to the array
+        parameter). The shape of the second dimension must be 4, as we are using bounding boxes with four corner points.
+        The points of the bounding box must be in clockwise order.The shape of the third dimension (the actual points
+        of each corner) should be 2. The first element in this dimension must be the x coordinate and the second
+        element must be the y coordinate.
+
+        :param xp: either numpy or cupy, depending on whether the code should run on CPU or GPU
+
+        :return: an array where all elements inside the bounding box get the value that is saved in `self.fill_value`.
+        """
+        if xp == cuda.cupy:
+            return self.create_mask_gpu(array, corners)
+        else:
+            return self.create_mask_cpu(array, corners)
```

Comment:
I'd prefer adding a newline to the end of the file (to prevent the small red icon from showing up:)

<img width="47" alt="image" src="https://user-images.githubusercontent.com/807665/49687285-20491680-fb01-11e8-8a18-8f315aa812ce.png">